components/scream/src/share/util/scream_tridiag.hpp

```diff
@@ -0,0 +1,564 @@
+             typename std::enable_if<TridiagDiag::rank == 1>::type* = 0,
+             typename std::enable_if<DataArray::rank == 1>::type* = 0,
+             impl::EnableIfCanUsePointer<TridiagDiag>* = 0,
+             impl::EnableIfCanUsePointer<DataArray>* = 0) {
+  const int nrow = d.extent_int(0);
+  assert( X.extent_int(0) == nrow);
+  assert(dl.extent_int(0) == nrow);
+  assert(du.extent_int(0) == nrow);
+  impl::thomas_a1x1(dl.data(), d.data(), du.data(), X.data(), nrow);
+}
+
+template <typename TridiagDiag, typename DataArray>
+KOKKOS_INLINE_FUNCTION
+void thomas (TridiagDiag dl, TridiagDiag d, TridiagDiag du, DataArray X,
+             typename std::enable_if<TridiagDiag::rank == 1>::type* = 0,
+             typename std::enable_if<DataArray::rank == 2>::type* = 0,
+             impl::EnableIfCanUsePointer<TridiagDiag>* = 0,
+             impl::EnableIfCanUsePointer<DataArray>* = 0) {
+  const int nrow = d.extent_int(0);
+  const int nrhs = X.extent_int(1);
+  assert( X.extent_int(0) == nrow);
+  assert(dl.extent_int(0) == nrow);
+  assert(du.extent_int(0) == nrow);
+  impl::thomas_a1xm(dl.data(), d.data(), du.data(), X.data(), nrow, nrhs);
+}
+
+template <typename TridiagDiag, typename DataArray>
+KOKKOS_INLINE_FUNCTION
+void thomas (TridiagDiag dl, TridiagDiag d, TridiagDiag du, DataArray X,
+             typename std::enable_if<TridiagDiag::rank == 2>::type* = 0,
+             typename std::enable_if<DataArray::rank == 2>::type* = 0,
+             impl::EnableIfCanUsePointer<TridiagDiag>* = 0,
+             impl::EnableIfCanUsePointer<DataArray>* = 0) {
+  const int nrow = d.extent_int(0);
+  const int nrhs = X.extent_int(1);
+  assert(X .extent_int(0) == nrow);
+  assert(dl.extent_int(0) == nrow);
+  assert(du.extent_int(0) == nrow);
+  assert(dl.extent_int(1) == nrhs);
+  assert(d .extent_int(1) == nrhs);
+  assert(du.extent_int(1) == nrhs);
+  impl::thomas_amxm(dl.data(), d.data(), du.data(), X.data(), nrow, nrhs);
+}
+
+// Cyclic reduction at the Kokkos team level. Any (thread, vector)
+// parameterization is intended to work.
+template <typename TeamMember, typename TridiagDiag, typename DataArray>
+KOKKOS_INLINE_FUNCTION
+void cr (const TeamMember& team,
+         TridiagDiag dl, TridiagDiag d, TridiagDiag du, DataArray X,
+         typename std::enable_if<TridiagDiag::rank == 1>::type* = 0,
+         typename std::enable_if<DataArray::rank == 1>::type* = 0) {
+  using Scalar = typename TridiagDiag::non_const_value_type;
+  const int nrow = d.extent_int(0);
+  assert(dl.extent_int(0) == nrow);
+  assert(d. extent_int(0) == nrow);
+  assert(du.extent_int(0) == nrow);
+  assert(X. extent_int(0) == nrow);
+  const int team_id = impl::get_thread_id_within_team(team);
+  const int nteam = impl::get_team_nthr(team);
+  int os = 1, stride;
+  // Go down reduction.
+  while ((stride = (os << 1)) < nrow) {
+    const int inc = stride*nteam;
+    for (int i = stride*team_id; i < nrow; i += inc) {
```

Comment:
Is the point of this manual for loop to avoid overhead from Kokkos' [TeamThread | ThreadVector]Range ||for?

Reply:
No. There are multiple issues, but I'll mention perhaps the most important, which is that the caller has configured (thread,vector) for her own application's benefit, and the solver may not do well with that config.